extension/src/prototypes/view/containers/KeyboardHandlers.ts

```diff
@@ -0,0 +1,138 @@
+import { useCallback, useRef, useState } from "react";
+
+import { useConstant, useWindowEvent } from "../../react-helpers";
+
+const directions = ["forward", "backward", "right", "left", "up", "down"] as const;
+
+type Direction = (typeof directions)[number];
+
+function isDirection(value: unknown): value is Direction {
+  return directions.includes(value as Direction);
+}
+
+const modes = ["sprint"] as const;
+
+type Mode = (typeof modes)[number];
+
+function isMode(value: unknown): value is Mode {
+  return modes.includes(value as Mode);
+}
+
+type KeyAssignments = Record<string, Direction | Mode>;
+
+const defaultKeyAssignments: KeyAssignments = {
+  KeyW: "forward",
+  KeyS: "backward",
+  KeyA: "left",
+  KeyD: "right",
+  Space: "up",
+  ControlLeft: "down",
+  ShiftLeft: "sprint",
+};
+
+export const KeyboardHandlers = () => {
+  const directionsRef = useRef<{
+    forward?: number;
+    backward?: number;
+    right?: number;
+    left?: number;
+    up?: number;
+    down?: number;
+  }>({});
+
+  const modesRef = useRef<{
+    sprint?: boolean;
+  }>({});
+
+  const state = useConstant(() => ({
+    time: Date.now(),
+    speed: 0,
+  }));
+  const camera = window.reearth?.camera;
+  const globeHeight = window.reearth?.scene?.getGlobeHeight();
+
+  const previousTime = state.time;
+  const currentTime = Date.now();
+  const deltaSeconds = (currentTime - previousTime) / 1000;
+  state.time = currentTime;
```

Comment:
Moving camera seems to be lagging... So you might need to use `requestAnimationFrame` instead of [the `usePrerender` which is used in the prototype](https://github.com/takram-design-engineering/plateau-view/blob/main/libs/cesium/src/KeyboardHandlers.ts#L150).

const renderFrame = () => {
  const currentTime = Date.now();
  const deltaSeconds = (currentTime - previousTime) / 1000;
  state.time = currentTime;
  checkOnKeyboard();
}

Reply:
I think the part has been updated.